### PR TITLE
Issue #2521: Django admin portfolio members section - [RJM]

### LINKED ISSUE #2521: Add portfolio member section to Django admin Portfolio view
### Issue description

As an analyst or superuser, I want to be able to look at the portfolio view and quickly know the human readable roles that members of the are assigned to. Permissions aren't easy to read so I want a quick way to see generally what they can do: Are they just a member, are they admin, readonly admin, a domain manager, can they make requests?

On django admin, we need to add the "Portfolio member" section to the Portfolio table and correctly show all users with their "role"/ human readable summary of their permissions (but not list the permissions themselves)

Note: We haven't solidified the member roles that will show to users, so consider this a first pass that will will update when member roles/permissions becomes more solidified. 

### Acceptance criteria

- [x] The portfolio view on django admin now has a Portfolio Member section that matches the [miro mockup](https://miro.com/app/board/uXjVKAEDJB4=/?moveToWidget=3458764593077028007&cot=14) 
- [ ] the rightmost column under show details shows any of the following View-only admin, Member, Domain manager, Domain requestor. If there is more than one of these that fits the users level of permissions& roles, show them as a comma separated list.  
- [x] Admins are not shown in the show details section and are shown in a way that matches the mockup
- [ ] If they have more permissions than just a normal member (such as they are a domain manager or can make/view requests) do not show the word "member" as that will be implied.

### Additional context
Note: that "view-only admin" is the newest name for read-only admin. as it ma
suggestion: User the @property field to add a field to the User table that provides their "portfolio_role_summary" which would be the list strings displayed in the second AC

bonus points if it is easy to have the permission/role labels stacked instead of as a comma separated list (not an AC though)
### Links to other issues

- blocked by #2520 
- blocks #2402

### PR DESCRIPTION
**Pull request by rachidatecs**
## Ticket

Resolves #2521 

## Changes

- 3 separate methods get admins, members, and members' summary
- Admin and members are hooked into the template as custom readonly fieldsets (reason for readonly documented in code)
- Members' summary is passed as extra context to the change form (reason for separation of this data is documented in code)
- All data items link to their change_form
- portfolio_role_summary is the brains behind the roles display logic. It will be revised but it establishes a good way to map permissions against what we want to display in admin.

## Context for reviewers

Add users with different roles AND-OR permissions to a portfolio and see the members' section populate.

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [x] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [x] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

**Comment by dave-kennedy-ecs** [comment]
When there are no admins in a portfolio, the messaging in DJA states, No Admins found.  When there are no members in a portfolio, the message is just {}.

**Comment by rachidatecs** [comment]
> No Admins found

Nice catch. Fixed.

**Comment by dave-kennedy-ecs** [comment]
I added the Rachid user as a Member and as an Admin to Hotel California.  On the Hotel California portfolio page, I see Rachid listed as an Admin, no members listed, and no table.  Is this the expected behavior?
image

**Comment by rachidatecs** [comment]
> I added the Rachid user as a Member and as an Admin to Hotel California. On the Hotel California portfolio page, I see Rachid listed as an Admin, no members listed, and no table. Is this the expected behavior? image

That's my understanding. We're separating admins from members.

**Comment by dave-kennedy-ecs** [comment]
> > I added the Rachid user as a Member and as an Admin to Hotel California. On the Hotel California portfolio page, I see Rachid listed as an Admin, no members listed, and no table. Is this the expected behavior? image
> 
> That's my understanding. We're separating admins from members.

cc: @abroddrick if there are only admins and no members, should the member table show?  Or is the table only intended to be shown for members?

**Comment by abroddrick** [comment]
> cc: @abroddrick if there are only admins and no members, should the member table show? Or is the table only intended to be shown for members?

@dave-kennedy-ecs good call out. Admins for sure shouldn't also appear in this table. I think for now, we should have the table still show if there are no non-admin members to make it clear. @Katherine-Osos may want some content to show there instead if there is no other members though. Katherine for the members table on the analysts Portfolio page, if there are admins and NO other members is it okay if we just have a blank table and then we see what analysts prefer later? Hiding the table seems like it may make it less clear that we are missing members, so if an empty table isn't ideal I would suggest we have just plain text that reads "No addition members in this portfolio"

**Review by dave-kennedy-ecs — APPROVED** [review]
Approving.
Please see the minor text change in Alysia's last comment before merging.

File: src/registrar/templates/django/admin/includes/detail_table_fieldset.html [hunk 1]
```diff
@@ -137,6 +137,16 @@
             {% endfor %}
             {% endwith %}
         </div>
+    {% elif field.field.name == "display_admins" %}
+        <div class="readonly">{{ field.contents|safe }}</div>
```
[the remark this answers is not in the document]
Comment (rachidatecs) [reply src/registrar/templates/django/admin/includes/detail_table_fieldset.html:141]: To convince DJA to parse HTML

File: src/registrar/admin.py [hunk 2]
```diff
@@ -2925,8 +2921,94 @@ class PortfolioAdmin(ListHeaderAdmin):
...
+            full_name = (
+                f"{member.first_name} {member.middle_name} {member.last_name}"
+                if member.middle_name
+                else f"{member.first_name} {member.last_name}"
+            )
+            member_details += "<tr>"
+            member_details += f"<td>{full_name}</td>"
+            member_details += f"<td>{member.title}</td>"
+            member_details += f"<td>{member.email}</td>"
+            member_details += f"<td>{member.phone}</td>"
```
[the remark this answers is not in the document]
Comment (rachidatecs) [reply src/registrar/admin.py:3097]: How does that work exactly?

File: src/registrar/admin.py [hunk 3]
```diff
@@ -3059,8 +3141,12 @@ def get_readonly_fields(self, request, obj=None):
...
 
     def change_view(self, request, object_id, form_url="", extra_context=None):
-        """Add related suborganizations and domain groups"""
-        extra_context = {"skip_additional_contact_info": True}
+        """Add related suborganizations and domain groups.
+        Add the summary for the portfolio members field (list of members that link to change_forms)."""
+        obj = self.get_object(request, object_id)
+        extra_context = extra_context or {}
+        extra_context["skip_additional_contact_info"] = True
+        extra_context["display_members_summary"] = self.display_members_summary(obj)
```
[the remark this answers is not in the document]
Comment (rachidatecs) [reply src/registrar/admin.py:3261]: No, I do. I do not want to display this as a field, I want to display it as an ADDITION to a field

File: src/registrar/templates/django/admin/includes/detail_table_fieldset.html [hunk 4]
```diff
@@ -137,6 +137,16 @@
...
+    {% elif field.field.name == "display_members" %}
+        <div class="readonly">
+            {% if display_members_summary %}
+                {{ display_members_summary }}
```
[the remark this answers is not in the document]
Comment (rachidatecs) [reply src/registrar/templates/django/admin/includes/detail_table_fieldset.html:145]: See above

**Pull request merged**